File: 2021/day-03/2021-03-1.py
#!/usr/bin/env python3
import sys
sys.path.append("../../libs/python")
import n0s1_aoc as n0s1

def bit_count(lines, i):
    return sum([((num >> i) & 1) for num in [int(i, 2) for i in lines]])

def get_power(lines):
    gamma = sum([(1 << i) if bit_count(lines, i) > len(lines)/2 else 0 for i in range(len(lines[0])-1,-1,-1)])
    return gamma * ((~gamma) & int("1" * len(lines[0]),2))

print(get_power(n0s1.file_to_strings("sample-2021-03.txt")))
print(get_power(n0s1.file_to_strings("input-2021-03.txt")))

File: 2021/day-03/2021-03-1.ex
defmodule Day3 do
  import N0S1_AOC

  def get_life_support(fname) do
    input = read_lines(fname)
    |> Enum.map(fn x ->
      String.split(x, "", trim: true)
      |> Enum.map(fn y ->
        String.to_integer(y)
      end)
    end)
    |> transpose

    threshold = length(hd(input)) / 2

    gamma = Enum.map(input, fn x ->
      if Enum.sum(x) >= threshold do 1 else 0 end
    end)

    epsilon = Enum.map(gamma, fn x -> (x - 1) * -1 end)

    bitlist_to_int(gamma) * bitlist_to_int(epsilon)
  end

  def main do
    IO.puts get_life_support("sample-2021-03.txt")
    IO.puts get_life_support("input-2021-03.txt")
  end

end
